common.blocks/i-bem-dom/__event-manager/_type/i-bem-dom__event-manager_type_bem.js

```diff
@@ -0,0 +1,231 @@
+/**
+ * @module i-bem-dom__event-manager_type_bem
+ */
+modules.define(
+    'i-bem-dom__event-manager_type_bem',
+    [
+        'i-bem__internal',
+        'inherit',
+        'identify',
+        'objects',
+        'jquery'
+    ],
+    function(
+        provide,
+        BEMINTERNAL,
+        inherit,
+        identify,
+        objects,
+        $) {
+
+var undef,
+    EVENT_PREFIX = '__bem__',
+
+    eventStorage = {},
+
+    buildEventName = function(e, entityName) {
+        return EVENT_PREFIX + entityName + (typeof e === 'object'? e.type : e);
+    },
+
+    buildBemEventManagerParams = function(ctx, defCtx, defSelector, defCls, getEntityCls) {
+        var res = {
+                entityCls : null,
+                ctx : defCtx,
+                selector : defSelector,
+                key : ''
+            };
+
+        if(ctx) {
+            var typeOfCtx = typeof ctx;
+
+            if(typeOfCtx === 'object' && ctx.__self) { // bem entity instance
+                res.ctx = ctx.domElem;
+                res.key = ctx._uniqId;
+                res.entityCls = ctx.__self;
+            } else if(typeOfCtx === 'string' || typeOfCtx === 'object' || typeOfCtx === 'function') {
+                var entityName, blockName, elemName, modName, modVal;
+                if(typeOfCtx === 'string') { // elem name
+                    blockName = defCls._blockName;
+                    elemName = ctx;
+                } else if(typeOfCtx === 'object') { // bem entity with optional mod val
+                    blockName = ctx.block?
+                        ctx.block.getName() :
+                        defCls._blockName;
+                    elemName = typeof ctx.elem === 'function'?
+                        ctx.elem.getName() :
+                        ctx.elem;
```

Comment:
Complexity over 9k.

Maybe like that? https://gist.github.com/zxqfox/7cbb43addf13473ddbba